### PR TITLE
V2 : primitives revisted

### LINKED ISSUE #626: V1 : Random artifacts on certain rounded rectangles
I've seen this manifest a number of times as I've been using OpenJSCAD. I finally got around to isolating a minimal example:

`cube({size: [50, 40, 10], radius: [3, 3, 0], fn: 15})`

Will produce a weird triangle to the side of the rounded cube. This artifact changes with different cube dimensions, radius, and fn quality.

## Expected Behavior

Just a rounded cube.

## Actual Behavior

A rounded cube with an additional non-manifold triangle to the side of it.

openjscad

## Steps to Reproduce the Problem

  1. Generate a model from:

```javascript
function main() {
  return cube({size: [50, 40, 10], radius: [3, 3, 0], fn: 15})
}
```

## Specifications

  - Version: 1.9.0
  - Platform: Chrome
  - Environment: Browser

### LINKED ISSUE #633: V1 : Add a warning/error message for cubes with dimensions of 0
## Expected Behavior
Give an error and continue to render fine when a cube is created with 1 or more dimensions equal to 0.

## Actual Behavior
Browser loads forever until a memory overflow (presumably). 

## Steps to Reproduce the Problem
1. Load the attached JSCAD file. 
2. See that it loads fine as it is.
3. Change the cube on Line 26 to be: `cube([1000, 0, 0])`.
4. Reload.
5. It loads forever until a memory overflow (presumably). 

## Specifications

  - Version: 1.9.0
  - Platform: Windows
  - Environment: browser
[Arm Rest Clamp.txt](https://github.com/jscad/OpenJSCAD.org/files/5004546/Arm.Rest.Clamp.txt)

### PR DESCRIPTION
**Pull request by z3dev**
These changes improve the checking of parameters (options) in the primitives.

NOTE: There are also changes to the tests suites to comply with coding standards.

Fixes #633 
Fixes #626 

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

**Comment by z3dev** [comment]
There's a lot of redundant code in the primitives, especially the redundant error messages/strings. I can see two ways to address this.

#1 - create a set of common functions to perform common checks, e.g. checkCenter()

#2 - keep the unique checks in the code, and create a set of common functions to throw common errors

Perferences anyone? Suggestions?

**Comment by SimonClark** [comment]
> There's a lot of redundant code in the primitives, especially the redundant error messages/strings. I can see two ways to address this.
> 
> 1. create a set of common functions to perform common checks, e.g. checkCenter()
> 2. keep the unique checks in the code, and create a set of common functions to throw common errors
> 
> Perferences anyone? Suggestions?

I think I'd prefer to see a set of options validator functions.  Make for more readable code imo

**Comment by z3dev** [comment]
@SimonClark thanks.

I implemented some common checks (actually more complete), and reworked the code for the primitives. Looks better, more readable, but I still don't like the redundant comments.

File: packages/modeling/src/primitives/arc.js [hunk 1]
```diff
@@ -27,13 +29,11 @@ const arc = (options) => {
...
   let { center, radius, startAngle, endAngle, makeTangent, segments } = Object.assign({}, defaults, options)
 
-  if (!Array.isArray(center)) throw new Error('center must be an array')
-  if (center.length < 2) throw new Error('center must contain X and Y values')
-
-  if (!Number.isFinite(radius)) throw new Error('radius must be a number')
-
-  if (startAngle < 0 || endAngle < 0) throw new Error('the start and end angles must be positive')
-  if (segments < 4) throw new Error('segments must be four or more')
+  if (!isArray(2, center)) throw new Error('center must be an array of X and Y values')
```
Comment (SimonClark) [inline packages/modeling/src/primitives/arc.js:32]: inconsistency in order of parameters here.  Should the thing being checked always be first?

File: packages/modeling/src/primitives/commonChecks.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// verify that the array has the given dimension, and contains Number values
+const isArray = (dimension, array) => {
+  if (Array.isArray(array) && array.length >= dimension) {
+    return array.every((n) => Number.isFinite(n))
+  }
+  return false
+}
+
+// verify that the value is a Number greater than the constant
+const isGT = (value, constant) => (Number.isFinite(value) && value > constant)
```
Comment (SimonClark) [inline packages/modeling/src/primitives/commonChecks.js:10]: I have a preference for isGreaterThan over isGT.  Minor thing, but it makes the code more like plain english.  Take it or leave it. :)

File: packages/modeling/src/primitives/commonChecks.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+// verify that the array has the given dimension, and contains Number values
+const isArray = (dimension, array) => {
```
Comment (SimonClark) [inline packages/modeling/src/primitives/commonChecks.js:2]: isArrayOfSize()?

File: packages/modeling/src/primitives/commonChecks.js (hunk 2 above)
Comment (z3dev) [reply packages/modeling/src/primitives/commonChecks.js:10]: I actually wrote the changes originally with isGreaterThan() but then I needed isGreaterThanOrEquals() which just seemed moronic for a simple check. It did look up abbreviations for these. :)

**Review by SimonClark — APPROVED** [review]

**Pull request merged**